amethyst_error/src/lib.rs

```diff
@@ -0,0 +1,402 @@
+//! Contains the `Error` type and company as used by Amethyst.
+//!
+//! **Note:** This type is not intended to be used outside of Amethyst.
+//! If you are integrating a crate from amethyst to use this, it is recommended that you treat this
+//! type as an opaque [`std::error::Error`].
+//!
+//! [`std::error::Error`]: https://doc.rust-lang.org/std/error/trait.Error.html
+
+#[cfg(all(feature = "backtrace", feature = "std"))]
+extern crate backtrace;
+
+use self::internal::Backtrace;
+use std::borrow::Cow;
+use std::error;
+use std::fmt;
+use std::result;
+
+/// Internal parts of `Error`.
+enum ErrorImpl {
+    Message(Cow<'static, str>),
+    Error(Box<dyn error::Error + Send + Sync>),
+}
+
+impl fmt::Debug for ErrorImpl {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            ErrorImpl::Message(ref message) => fmt::Debug::fmt(message, fmt),
+            ErrorImpl::Error(ref e) => fmt::Debug::fmt(e, fmt),
+        }
+    }
+}
+
+impl From<&'static str> for ErrorImpl {
```

Comment:
I definitely see the motivation for this, but I think we should discourage using strings anywhere in the engine.

Reply:
I'm not sure why static strings should be discouraged?

Reply:
Discourages low effort error handling. Strings should be embedded at the lower leaf level of errors, not on the high level type.

I agree with @torkleyy I would like to see the `Message` part dropped imo.

Reply:
This can certainly be done once we've refactored all instances where we use them, but not before. If like to punt that as out of scope for now since it would be a significant undertaking.